crhchttp/config.go

```diff
@@ -6,19 +6,134 @@
+// The fields within the struct provide information about the access token that is used to authenticate service account requests.
+type ServiceAccountToken struct {
+	AccessToken      string `json:"access_token"`
+	ExpiresIn        int    `json:"expires_in"`
+	RefreshExpiresIn int    `json:"refresh_expires_in"`
+	TokenType        string `json:"token_type"`
+	NotBeforePolicy  int    `json:"not-before-policy"`
+	Scope            string `json:"scope"`
+}
+
+const serviceaccount = metricscfgv1beta1.ServiceAccount
+
+// AuthError represents a client error returned when authenticating client credentials.
+// It includes the HTTP status code, an error type specific to the authentication process and a desriptive error message.
+type AuthError struct {
+	StatusCode  int
+	ErrorType   string
+	Description string
+}
+
+// Error returns a formated error string combining the HTTP status, error type and description.
+func (e *AuthError) Error() string {
+	return fmt.Sprintf("HTTP Status: %d, Error: %s, Description: %s", e.StatusCode, e.ErrorType, e.Description)
+}
+
+func (ac *AuthConfig) GetAccessToken(cxt context.Context, tokenURL string) error {
+	if ac.Authentication != serviceaccount {
+		return nil
+	}
+
+	log := log.WithName("GetAccessToken")
+
+	// Prepare the POST data
+	data := url.Values{}
+	data.Set("client_id", ac.ServiceAccountData.ClientID)
+	data.Set("client_secret", ac.ServiceAccountData.ClientSecret)
+	data.Set("grant_type", ac.ServiceAccountData.GrantType)
+
+	// // Making the HTTP POST request
+	cxt, cancel := context.WithTimeout(cxt, 5*time.Second)
+	defer cancel()
+	req, err := http.NewRequestWithContext(cxt, http.MethodPost, tokenURL, bytes.NewBufferString(data.Encode()))
+	if err != nil {
+		return fmt.Errorf("failed to construct HTTP request: %w", err)
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to make HTTP request to acquire token: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	// Handle client errors and invalid requests
+	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+		var errResponse struct {
+			Error            string `json:"error"`
+			ErrorDescription string `json:"error_description"`
+		}
+
+		if err := json.Unmarshal([]byte(body), &errResponse); err != nil {
+			// return generic error with addiotional context if unmarshalling fails
+			return fmt.Errorf("status: %d, failed to unmarshal error response: %w", resp.StatusCode, err)
+		}
+
+		return &AuthError{
+			StatusCode:  resp.StatusCode,
+			ErrorType:   errResponse.Error,
+			Description: errResponse.ErrorDescription,
+		}
+	}
+
+	var result ServiceAccountToken
+	if err := json.Unmarshal([]byte(body), &result); err != nil {
+		return fmt.Errorf("failed to unmarshal response: %w", err)
+	}
```

Comment:
A few things,
1. let's use `json.NewDecoder(resp.Body).Decode(&target)` instead of `io.ReadAll` (I'm not sure where I had gotten io.readall)
2. let's go thru the error pathway for every response that is not a 200. I think we can assume that if the response is not a 200, then the response is not what we need.
3. Let's also use the `AuthError` directly instead of creating a one off `errResponse` struct.
```
type AuthError struct {
	StatusCode       int
	ErrorName        string `json:"error"`
	ErrorDescription string `json:"error_description"`
}

// Error returns a formated error string combining the HTTP status, error type and description.
func (e AuthError) Error() string {
	return fmt.Sprintf("HTTP Status: %d, Error: %s, Description: %s", e.StatusCode, e.ErrorName, e.ErrorDescription)
}
```

```suggestion
	// Handle client errors and invalid requests
	if resp.StatusCode != 200 {
		errResponse := AuthError{StatusCode: resp.StatusCode}

		if err := json.NewDecoder(resp.Body).Decode(&errResponse); err != nil {
			// return generic error with addiotional context if unmarshalling fails
			return fmt.Errorf("status: %d, failed to unmarshal error response: %w", resp.StatusCode, err)
		}
		return errResponse
	}

	var result ServiceAccountToken
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return fmt.Errorf("failed to unmarshal response: %w", err)
	}
```